lib/srv/db/postgres/engine.go

```diff
@@ -0,0 +1,454 @@
+		sessionCtx.Server.GetURI(),
+		sessionCtx.Server.GetRegion(),
+		sessionCtx.DatabaseUser,
+		e.Credentials)
+}
+
+// getTLSConfig builds the client TLS configuration for the session.
+//
+// For RDS/Aurora, the config must contain RDS root certificate as a trusted
+// authority. For onprem we generate a client certificate signed by the host
+// CA used to authenticate.
+func (e *Engine) getTLSConfig(ctx context.Context, sessionCtx *session.Context) (*tls.Config, error) {
+	addr, err := utils.ParseAddr(sessionCtx.Server.GetURI())
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	tlsConfig := &tls.Config{
+		ServerName: addr.Host(),
+		RootCAs:    x509.NewCertPool(),
+	}
+	// Add CA certificate to the trusted pool if it's present, e.g. when
+	// connecting to RDS/Aurora which require AWS CA.
+	if len(sessionCtx.Server.GetCA()) != 0 {
+		if !tlsConfig.RootCAs.AppendCertsFromPEM(sessionCtx.Server.GetCA()) {
+			return nil, trace.BadParameter("failed to append CA certificate to the pool")
+		}
+	} else if sessionCtx.Server.IsAWS() {
+		if rdsCA, ok := e.RDSCACerts[sessionCtx.Server.GetRegion()]; ok {
+			if !tlsConfig.RootCAs.AppendCertsFromPEM(rdsCA) {
+				return nil, trace.BadParameter("failed to append CA certificate to the pool")
+			}
+		} else {
+			e.Log.Warnf("No RDS CA certificate for %v.", sessionCtx.Server)
+		}
+	}
+	// RDS/Aurora auth is done via an auth token so don't generate a client
+	// certificate and exit here.
+	if sessionCtx.Server.IsAWS() {
+		return tlsConfig, nil
+	}
+	// Otherwise, when connecting to an onprem database, generate a client
+	// certificate. The database instance should be configured with
+	// Teleport's CA obtained with 'tctl auth sign --type=db'.
+	cert, cas, err := e.getClientCert(ctx, sessionCtx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	tlsConfig.Certificates = []tls.Certificate{*cert}
+	for _, ca := range cas {
+		if !tlsConfig.RootCAs.AppendCertsFromPEM(ca) {
+			return nil, trace.BadParameter("failed to append CA certificate to the pool")
+		}
+	}
+	return tlsConfig, nil
+}
+
+// getClientCert signs an ephemeral client certificate used by this
+// server to authenticate with the database instance.
+func (e *Engine) getClientCert(ctx context.Context, sessionCtx *session.Context) (cert *tls.Certificate, cas [][]byte, err error) {
+	privateBytes, _, err := native.GenerateKeyPair("")
+	if err != nil {
+		return nil, nil, trace.Wrap(err)
+	}
+	// Postgres requires the database username to be encoded as a common
+	// name in the client certificate.
+	subject := pkix.Name{CommonName: sessionCtx.DatabaseUser}
+	csr, err := tlsca.GenerateCertificateRequestPEM(subject, privateBytes)
+	if err != nil {
+		return nil, nil, trace.Wrap(err)
+	}
+	e.Log.Debugf("Generating client certificate for %s.", sessionCtx)
+	resp, err := e.AuthClient.GenerateDatabaseCert(ctx, &proto.DatabaseCertRequest{
+		CSR: csr,
+		TTL: proto.Duration(sessionCtx.Identity.Expires.Sub(e.Clock.Now())),
+	})
```

Comment:
this is expensive, but cache-able
maybe leave a TODO to cache DB certs, since this PR is already big

Reply:
Yeah, for sure, was thinking about caching as well. I'll leave a TODO item.